reuse.go

```diff
@@ -0,0 +1,190 @@
+package libp2pquic
+
+import (
+	"errors"
+	"net"
+	"sync"
+
+	srcs "github.com/lnykww/go-src-select"
+)
+
+type ReuseConn struct {
+	net.PacketConn
+	lock sync.Mutex
+	ref  int
+}
+
+func NewReuseConn(conn net.PacketConn) *ReuseConn {
+	return &ReuseConn{
+		PacketConn: conn,
+		ref:        1,
+		lock:       sync.Mutex{},
+	}
+}
+
+func (rc *ReuseConn) Ref() error {
+	rc.lock.Lock()
+	defer rc.lock.Unlock()
+	if rc.ref == 0 {
+		return errors.New("conn closed")
+	}
+	rc.ref++
+	return nil
+}
+
+func (rc *ReuseConn) Close() error {
+	rc.lock.Lock()
+	defer rc.lock.Unlock()
+	var err error
+	switch rc.ref {
+	case 0: // cloesd, just return
+		return nil
+	case 1: // no reference, close the conn
+		err = rc.PacketConn.Close()
+	}
+	rc.ref--
+	return err
+}
+
+type Reuse struct {
+	lock           sync.Mutex
+	unicast        map[string]map[int]net.PacketConn
+	unspecific     []net.PacketConn
+	connGlobal     net.PacketConn
+	connGlobalOnce sync.Once
+}
+
+func NewReuse() *Reuse {
+	return &Reuse{
+		unicast:    make(map[string]map[int]net.PacketConn),
+		unspecific: make([]net.PacketConn, 0),
+	}
+}
+
+// getConnGlobal get the global random port socket, if not exist, create
+// it first.
+func (r *Reuse) getConnGlobal(network string) (net.PacketConn, error) {
+	var err error
+	r.connGlobalOnce.Do(func() {
+		var addr *net.UDPAddr
+		var conn net.PacketConn
+		var host string
+		switch network {
+		case "udp4":
+			host = "0.0.0.0:0"
+		case "udp6":
+			host = "[::]:0"
+		}
+		addr, err = net.ResolveUDPAddr(network, host)
+		if err != nil {
+			return
+		}
+		conn, err = net.ListenUDP(network, addr)
+		if err != nil {
+			return
+		}
+
+		r.connGlobal = NewReuseConn(conn)
+	})
+	if r.connGlobal == nil && err == nil {
+		err = errors.New("global socket init not done")
+	}
+	return r.connGlobal, err
+}
+
+// rueseConn Assertion the type of the conn is ReuseConn and inc the ref
+func (r *Reuse) reuseConn(conn net.PacketConn) error {
+	reuseConn, ok := conn.(*ReuseConn)
+	if !ok {
+		panic("type ReuseConn Assert failed: something wrong!")
+	}
+	return reuseConn.Ref()
+}
+
+func (r *Reuse) dial(network string, raddr *net.UDPAddr) (net.PacketConn, error) {
+	// Find the source address which kernel use
+	sip, err := srcs.Select(raddr.IP)
+	if err != nil {
+		return r.getConnGlobal(network)
+	}
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	// If we has a listener on this address, use it to dial
+	if c, ok := r.unicast[sip.String()]; ok {
+		for _, v := range c {
+			return v, nil
+		}
+	}
+
+	if len(r.unspecific) != 0 {
+		return r.unspecific[0], nil
+	}
+
+	return r.getConnGlobal(network)
+}
+
+func (r *Reuse) Dial(network string, raddr *net.UDPAddr) (net.PacketConn, error) {
+	conn, err := r.dial(network, raddr)
+	if err != nil {
+		return nil, err
+	}
+	// we are reuse a conn, reference it
+	if err = r.reuseConn(conn); err != nil {
```

Comment:
* I believe this will double-ref if we end up creating a new global connection.
* We probably need a loop (either inside dial out outside). Otherwise, we could try to use the connection but fail because some other thread decided it didn't want it. In that case, we should just open a new connection.

Reply:
> * I believe this will double-ref if we end up creating a new global connection.

I have found this problem before. But we create a listen socket as a global socket, 
it means that if we don't want to use it, we need close it ourself in this module, 
rather than the user in other module. So when it was created, the reference should be 1. 

> * We probably need a loop (either inside dial out outside). Otherwise, we could try to use the connection but fail because some other thread decided it didn't want it. In that case, we should just open a new connection.

Ok, i will add a retry for that.

Reply:
> I have found this problem before. But we create a listen socket as a global socket,
it means that if we don't want to use it, we need close it ourself in this module,
rather than the user in other module. 

My thinking was that, when the last connection using the global connection closes, we could close the global connection (reopening a new one when necessary).

Reply:
> My thinking was that, when the last connection using the global connection closes, we could close the global connection (reopening a new one when necessary).

This may cause us to create new socket frequently.

Reply:
If we're repeatedly closing all connections. Also note: this will only happen if we aren't listening on a relevant ip address.

My worry here is that we're _leaking_ a socket. In theory, a user should be able to spin up/down new libp2p instances without leaking anything.

Reply:
Yes, you are right. It does not happen very often. I will do that.